runners/jet/src/main/java/org/apache/beam/runners/jet/Utils.java

```diff
@@ -0,0 +1,249 @@
+  }
+
+  static boolean usesStateOrTimers(AppliedPTransform<?, ?, ?> appliedTransform) {
+    try {
+      return ParDoTranslation.usesStateOrTimers(appliedTransform);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static DoFn<?, ?> getDoFn(AppliedPTransform<?, ?, ?> appliedTransform) {
+    try {
+      DoFn<?, ?> doFn = ParDoTranslation.getDoFn(appliedTransform);
+      if (DoFnSignatures.signatureForDoFn(doFn).processElement().isSplittable()) {
+        throw new IllegalStateException(
+            "Not expected to directly translate splittable DoFn, should have been overridden: "
+                + doFn); // todo
+      }
+      return doFn;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static WindowingStrategy<?, ?> getWindowingStrategy(AppliedPTransform<?, ?, ?> appliedTransform) {
+    // assume that the windowing strategy is the same for all outputs
+
+    Map<TupleTag<?>, PValue> outputs = getOutputs(appliedTransform);
+
+    if (outputs == null || outputs.isEmpty()) {
+      throw new IllegalStateException("No outputs defined.");
+    }
+
+    PValue taggedValue = outputs.entrySet().iterator().next().getValue();
+    checkState(
+        taggedValue instanceof PCollection,
+        "Within ParDo, got a non-PCollection output %s of type %s",
+        taggedValue,
+        taggedValue.getClass().getSimpleName());
+    PCollection<?> coll = (PCollection<?>) taggedValue;
+    return coll.getWindowingStrategy();
+  }
+
+  /**
+   * Assigns the {@code list} to {@code count} sublists in a round-robin fashion. One call returns
+   * the {@code index}-th sublist.
+   *
+   * <p>For example, for a 7-element list where {@code count == 3}, it would respectively return for
+   * indices 0..2:
+   *
+   * <pre>
+   *   0, 3, 6
+   *   1, 4
+   *   2, 5
+   * </pre>
+   */
+  public static <T> List<T> roundRobinSubList(List<T> list, int index, int count) {
+    if (index < 0 || index >= count) {
+      throw new IllegalArgumentException("index=" + index + ", count=" + count);
+    }
+    return IntStream.range(0, list.size())
+        .filter(i -> i % count == index)
+        .mapToObj(list::get)
+        .collect(toList());
+  }
+
+  /** Returns a deep clone of an object by serializing and deserializing it (ser-de). */
+  @SuppressWarnings("unchecked")
+  public static <T> T serde(T object) {
+    try {
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      ObjectOutputStream oos = new ObjectOutputStream(baos);
+      oos.writeObject(object);
+      oos.close();
+      byte[] byteData = baos.toByteArray();
+      ByteArrayInputStream bais = new ByteArrayInputStream(byteData);
+      return (T) new ObjectInputStream(bais).readObject();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static <T> byte[] encodeWindowedValue(
+      WindowedValue<T> windowedValue, Coder coder, ByteArrayOutputStream baos) {
```

Comment:
Possibly you want to use `CoderUtils.encode(coder, value)`.

Reply:
True, thank you, will do so in the next commit.